### ISSUE TITLE
Use handler-bind instead of handler-case

### ISSUE THREAD

**Issue by orthecreedence** [orthecreedence/blackbird#8]
doesn't unwind the stack when handling errors

**Comment by orthecreedence** [comment]
Ok, please see https://github.com/orthecreedence/cl-async/issues/108. I decided to keep event-cb because it is used to catch events from libuv on a per-connection basis, but other than the conditions created by libuv itself, all conditions are now allowed to bubble up without being touched or passed into event-cb (and enter the debugger) if `:catch-app-errors nil` is passed (which is the default). If `:catch-app-errors t` is given, ALL errors are caught and passed to the function `:caught-errors` (given on startup).

There are more details in the cl-async issue. Please give it a shot, and let me know if you need changes!

**Comment by capitaomorte** [comment]
I tried loading the very bleeding edge of cl-async and its dependencies and after some effort, I managed to do it.

I was trying to run this very basic example

```
(with-event-loop ()
      (tcp-connect "www.google.com" 80
                   (lambda (socket data)
                     (when t
                       (format t "Done processing!~%")
                       (close-socket socket)))  ; close the socket if done processing
                   (lambda (error) (format t "Ooops an error!"))
                   :data (format nil "GET /~c~c" #\return #\newline)))
```

and make variations on it so I could check if error handling is now cool and if I can make my `handler-bind`s around this and establish restarts near the "done processing" and all that.

But I crashed into this, and didn't debug any further, sorry. Perhaps you can help me?

```
The alien function "uv_loop_size" is undefined.
   [Condition of type SB-KERNEL::UNDEFINED-ALIEN-FUNCTION-ERROR]

Restarts:
 0: [RETRY] Retry SLY mREPL evaluation request.
 1: [*ABORT] Return to SLY's top level.
 2: [ABORT] Abort thread (#<THREAD "sly-channel-1-mrepl-remote-1" RUNNING {1003483CA3}>)

Backtrace:
  0: ("undefined function")
  1: (LIBUV:UV-LOOP-SIZE)
  2: (SB-INT:SIMPLE-EVAL-IN-LEXENV (WITH-EVENT-LOOP NIL (TCP-CONNECT "www.google.com" 80 (LAMBDA # #) (LAMBDA NIL #) :DATA ...)) #<NULL-LEXENV>)
  3: (EVAL (WITH-EVENT-LOOP NIL (TCP-CONNECT "www.google.com" 80 (LAMBDA # #) (LAMBDA NIL #) :DATA ...)))
  4: ((LAMBDA NIL :IN SLYNK-MREPL::MREPL-EVAL-1))
```

**Comment by orthecreedence** [comment]
Seems like an outdated version of libuv. Can you confirm you have libuv >= 1.0.0?

**Comment by capitaomorte** [comment]
No I can't, because I don't :-). Homebrew on Mac OSX gives me libuv 0.10.21. Building from head right now, will be back with some more info soon. Shouldn't cffi and such warn early when compiling bindings?

**Comment by capitaomorte** [comment]
After much effort (had to trash quicklisp for some reason), my little example now yields another error. Using a pristine quicklisp, sbcl 1.2.2, master cl-async, cl-libuv and fast-io

```
#<DNS-ERROR -1: the addrinfo->ai_addr object was null (stinks of a memory alignment issue) {10060C64A3}>
   [Condition of type DNS-ERROR]
```

But we already have an illustration of the problems that I think still exist with cl-async.
I've noticed already that `:catch-app-errors` set to nil doesn't catch this error. Is this a non-app error in your view? But why is it being caught? I think this is the fundamental question, btw. So that the loop can go on and process other events? We can do this behaviour in production (with a default handler that just logs the error to some file). In development let it block the thread and all other pending events, so that the memory alignment smells can be analysed from their origin. As it stands I had to resignal it:
```lisp
(with-event-loop (:catch-app-errors nil)
      (tcp-connect "www.google.com" 80
                   (lambda (socket data)
                     (format t "Done processing!~%")
                     (close-socket socket))
                   (lambda (error) (error error))
                   :data (format nil "GET /~c~c" #\return #\newline)))
```


BIG UPDATE. I didn't read the doc or you didn't write it yet, but I've noticed that `:caught-errors` does what I wish and apparently the error originates from the c callback `dns-cb`. Much better.  I still wonder about some other parts, but so far so god.

**Comment by orthecreedence** [comment]
You're right, that memory alignment error should not be getting passed to the event-cb. That is a special case I forgot about when doing the conversion. I'll review each of the subsystems' C callbacks today and make sure that non-libuv conditions get thrown instead of passed to event-cb.

**Comment by orthecreedence** [comment]
Oh also, I'll spin up my mac (or bsd, similar) VM today and try to track down that actual DNS error. I might just cave in the next few days and rebuild cl-libuv via cffi-groveler so this type of junk doesn't happen anymore.

**Comment by orthecreedence** [comment]
I fixed that DNS throwing issue (the error is now thrown) and checked all the other systems to make sure conditions are being treated properly. Here's the breakdown: any libuv status that comes through a libuv callback that is not a "success" will have an event condition created for it and be passed to the attached event-cb. Any libuv status that comes back from calling a libuv function directly (such as writing to a socket) that is not a "success" will have an error *thrown* for it. If an error is not thrown, then it's a bug in cl-async. Any errors that occur in non-cl-async code (but were spawned via cl-async) are not touched by cl-async at all, unless `:catch-app-errors t` is passed.

In other words, the only thing event-cb is used for now is things like "socket timed out" or "dns not found" etc. Operational errors ("writing to socket returned an error" et al) throw errors. All thrown errors are untouched.

---

However, I still have yet to track down the root cause of the DNS error you're seeing. I tried `dns-lookup` on sbcl on a BSD machine and it just hangs, no errors and no output. So that's broken there for sure. Didn't have time to keep digging. I'll try spinning up my OSx VM and give it a shot (if it works, the VM is really finnicky and I don't have an extra mac laying around so wish me luck).

**Comment by capitaomorte** [comment]
> Here's the breakdown: any libuv status that comes through a libuv callback that is not a "success" will have an event condition created for it and be passed to the attached event-cb.

This part I almost understand, untill the part where you talk about "passing it to the attached event-cb". Why not let it bubble up to top level? In other words, why not make it behave like the next category of unexpected behaviour, which I'll comment on shortly?

> Any libuv status that comes back from calling a libuv function directly (such as writing to a socket) that is not a "success" will have an error thrown for it.

This is quite OK. But is the caught by some `handler-case` or `handler-bind` in some conditions? In a production scenario I do want these errors to be handled. Should my client code do that? Via a argument somewhere like:

```lisp
(in-package :capitaomorte)

(cl-async:some-good-function #'process-success
   :error-handling #'log-error))
```
, or by wrapping my own handler-case around some `cl-async::fiasco` condition class? 

```lisp
(in-package :capitaomorte)

(handler-case 
   (cl-async:some-good-function #'process-success)
  (cl-async:fiasco (f) (log-error f))
```

Or even better, if `cl-async` provides some cool restarts that I can use, I'll do

```lisp
(in-package :capitaomorte)

(handler-bind
    ((cl-async:fiasco (lambda (f)
                        (when (find-restart 'cl-async:log-it-and-continue)
                          (invoke-restart 'cl-async:log-it-and-continue))
                        ;; must be some more serious fiasco
                        ;; that doesn't provide this restart so
                        ;; let it bubble up to someone that knows how
                        ;; to handle it
                        ))
     ;; In this scenario, I imagine that this one may be thrown 
     ;; by PROCESS-SUCCESS. In the others I would need 
     ;; similar
     (capitaomorte:usual-f*ckup
       (lambda (f)
         (if (capitaomorte:production-mode-p)
             (invoke-restart 'capitaomorte:muffle-it)))))
  (cl-async:some-good-function #'process-success))                        
```

> I'll try spinning up my OSx VM and give it a shot (if it works, the
> VM is really finnicky and I don't have an extra mac laying around so
> wish me luck).

I can help track this down too and give my socket-fu a polish.

**Comment by orthecreedence** [comment]
> This part I almost understand, untill the part where you talk about "passing it to the attached event-cb". Why not let it bubble up to top level?

My reasoning for this is to be able to handle events differently on a per-object basis. For instance

```lisp
(as:with-event-loop ()
  (as:tcp-connect
    "www.google.com" 80
    (lambda (sock data)
      (my-app:do-stuff data))
    (lambda (ev)
      (when (typep ev 'as:socket-eof)
        (my-app:we-got-booted (as:tcp-socket ev))))))
```

I like this because a socket hanging up isn't worth exiting the event loop, IMO. In order to achieve the same thing with `as:socket-eof` being thrown, we'd have to do this:

```lisp
(handler-bind
    ((as:socket-eof
       (lambda (ev)
         (my-app:we-got-booted (as:tcp-socket ev))
         (let ((restart (find-restart 'cl-async:continue-event-loop ev)))
           (when restart (invoke-restart restart))))))
  (as:with-event-loop ()
    (as:tcp-connect
      "www.google.com" 80
      (lambda (sock data)
        (my-app:do-stuff data)))))
```

This seems like a lot of boilerplate and mental overhead to handle something as simple as a socket EOFing, which is fairly commonplace. And consider this scenario:

```lisp
(as:with-event-loop ()
  ;; analytics are nice but not critical
  (as:tcp-connect
    "my-analytics.com" 80
    (lambda (sock data) ...)
    (lambda (e)
      (when (typep e 'as:socket-refused)
        (format t "call to analytics server failed, no big deal ~a~%" e))))
  ;; we need this
  (as:tcp-connect
    "launch-the-missle.net" 80
    (lambda (sock data) ...)
    (lambda (e)
      (when (typep e 'as:socket-refused)
        (error "Nuke launch failed!")))))
```

Would this even have a direct translation in a `handler-bind` form? We can't wrap the individual forms with handler-bind because the callbacks are invoked under a different stack than the calling function, so by the time the error is thrown the stack has completely unwound and our error handlers are gone. So we'd have to wrap the outside of the `with-event-loop` with a big handler-bind form and somehow distinguish between our analytics socket and our nuke socket if an error occurs.

On top of this, from my understanding (helped quite a bit by talking to you) one of the main benefits for letting errors bubble up is to a) get backtraces and b) drop into the debugger. A backtrace here is useless because it's only a few levels deep under `start-event-loop`, and I'm not convinced debugging is the right option either because it seems like the only two viable options here are to either exit the event loop (which happens automatically if that socket was the only event left to process) or to pass the condition to event-cb and keep processing the loop (which is how things currently work).

In other words, I'm just not sure it would be useful to completely abolish event-cb in cases where libuv is actually handing us events having to do with the corresponding objects we're operating on. And in doing so, it may create a lot more overhead by forcing all error handling to the outside of the event loop (we can't use handler-case or handler-bind around async operations so we need to put them outside the event loop) as well as making it a challenge to identify which errors are occuring on which objects.

>This is quite OK. But is the caught by some handler-case or handler-bind in some conditions? In a production scenario I do want these errors to be handled. Should my client code do that?

Yes, this would be the calling code's responsibility. All errors thrown *should* have the base `cl-async:event-error` so you could do something like:

```lisp
(handler-case
  (as:with-event-loop ()
    (my-app:do-things-that-might-throw-errors ...))
  (as:event-error (e)
    (my-app:log-error e)))
```

The one exception I know of is the DNS error you found, which is actually more of a panic than anything else, and is thrown with the type `cl:error`, although I am open to changing this to another error type. This is thrown *after* a DNS request comes back (async) though, so it would have to be caught *outside* the event loop.

As mentioned over in https://github.com/orthecreedence/cl-async/issues/108, `:catch-app-errors` will catch __all__ errors that happen in any code running under cl-async and pass them to `:caught-errors` *with the exception* of errors that happen when calling a libuv function directly, in which case cl-async doesn't catch the error at all. This can be something like writing to a socket that's closed, opening a server on a port that's already bound, etc. These errors are thrown and are not handled by cl-async at all, even with `:catch-app-errors t`. 

In fact, I'm starting to wonder if `:catch-app-errors t` is even useful, because it can now be completely mimicked by doing:

```lisp
(handler-bind
    ((error (lambda (c)
              (let ((res (find-restart 'cl-async-util:continue-event-loop c)))
                (when res (invoke-restart res))))))
  (as:with-event-loop (:catch-app-errors nil
                       :caught-errors
                         (lambda (e)
                           (format t "- caught: ~a~%" e)))
    (as:with-delay ()
      (format t "x is ~a~%" (+ 4 'five)))))
```

`continue-event-loop` passes the caught error to `:caught-errors` and aborts the current callback but keeps the loop running. Thus, you can continue the loop on a per-condition basis. this is all undocumented so far, and will probably remain that way until things solidify over the next few weeks.

**Comment by capitaomorte** [comment]
> This seems like a lot of boilerplate and mental overhead to handle something as simple as a socket EOFing, which is fairly commonplace. 

The "boilerplate" argument doesn't hold much for me in common lisp, because macros. `cl-async` could well provide a `restarting-on-hangup` macro that does that, or a slightly more generally named macro with more flexibility and some default behaviour.

Also note that I'm not saying that that parameter should be extinct, only optional.

> Would this even have a direct translation in a handler-bind form? We can't wrap the individual forms with handler-bind because the callbacks are invoked under a different stack than the calling function

I see your point, indeed. The :event-cb is useful there indeed, and now I understand the intrinsic need for "attaching" handlers. But why not let it have `handler-bind` semantics? I.e, if it doesn't invoke a restart or throw to some some exported `cl:event-handling-done` tag, then the error does bubble up. You capture the lambda that the user passes into `:event-cb` and simply funcall it from a `handler-bind` inside `cl-async`, wrapped inside a `cl:event-handling-done` tag. THis is unfortunately backward incompatible to existing code, so some solution would have to be found for that.

Tangentially, I personally also hate the fact that I have to use `typep`. I'd rather use CLOS for that, and declaratively define

```elisp
(defmethod my-app:handle-event ((event as:socket-refused))
   (if (string= (url-of event)) "analytics") ... ...))
```

And then pass `#'my-app:handle-event` as the error handler. CLOS should do the right thing.

Actually `cl-async` could provide a basic `cl-async:handle-error` generic function with some preloaded methods, and export it. The export means client code can add methods to it. I'll explain more below.

> main benefits for letting errors bubble up is to a) get backtraces and b) drop into the debugger

There is also the benefit that any restarts either `cl-async`'s or mine established along the way can be called from any handler-bind

> In other words, I'm just not sure it would be useful to completely abolish event-cb in cases where libuv is actually handing us events having to do with the corresponding objects we're operating on. And in doing so, it may create a lot more overhead by forcing all error handling to the outside of the event loop (we can't use handler-case or handler-bind around async operations so we need to put them outside the event loop) as well as making it a challenge to identify which errors are occuring on which objects.

Yes, this argument is made. Don't abolish it. Optionally give it `handler-bind` semantics. To ensure backward compatibility, make `handler-case` semantics the default. 


>| Yes, this would be the calling code's responsibility. All errors thrown should have the base cl-async:event-error so you could do something like:

I assume you meant "all errors thrown by `cl-async`". In that case, it is more important that they all eventually provide a `cl-async-util:continue-event-loop` restart than their exact type, cause that's really what we're looking for here.

> The one exception I know of is the DNS error you found, which is actually more of a panic than anything else, and is thrown with the type cl:error, although I am open to changing this to another error type. 

I guess. But if someday you feel that `cl-async` should delegate to some other library that uses some other error class, then don't resort to catching, changing the error class and resignalling, use restarts instead.

> In fact, I'm starting to wonder if :catch-app-errors t is even useful, because it can now be completely mimicked by doing:

Yes exactly. But I don't find it horrible, keep it. It's nice to pass it the value returned by some function `(my-app:production-mode-p)`.

Or, I'd make it a generalized boolean and get rid of `:caught-errors` entirely (or rather deprecate it). So if one passes `nil` to `:catch-app-errors`, *everything* bubbles up. If one passes `t` to it everything is handled through the `as:handle-error` generic function, which has some nice built-in methods that catch only `cl-async` errors and don't mess with user-specific errors at all. If one passes it a lambda, call that lambda with the error class. That function is resposible for handling all the errors. The user has the choice to add methods to `as:handle-error` or call it from his/her lambda.

This very same behaviour would go for the fourth argument of functions such as `as:tcp-connect` and overrides the behaviour for `:catch-app-errors`.

So to summarize, here's what I'd like to see:

* make the fourth argument of functions such as `as:tcp-connect` optional. `nil` means everything bubbles up. t means use `as:handle-error`, a new generic and a lambda is a lambda as usual, but called with `handler-bind` semantics, i.e. if it doesn't throw or invoke restarts the error continues its journey.

* add an exported catch tag of `cl:event-handling-done` alongside the `cl-async-util:continue-event-loop` that you apparently already provide. Export both.

* Add a `as:handle-error` generic function and define some methods for it. Export that.

* make the `:catch-app-errors` argument the same semantics as the previous item. it's only called when a specific request doesn't have an error handler attached, which is the point you made earlier. 

* `error-catcher` is deprecated.

How backward incompatible is all this? The new `handler-bind` semantics certainly break some existing code. The optionality and generalized booleaness of `:catch-app-errors` and the fourth arg to `as:tcp-connect` are less of a problem, but would still need adjustments.

Sorry for the long-windedness and What do you think?

**Comment by orthecreedence** [comment]
Sorry in advance for the great wall of text.

> The "boilerplate" argument doesn't hold much for me in common lisp

True =].

> But why not let it have handler-bind semantics? I.e, if it doesn't invoke a restart or throw to some some exported cl:event-handling-done tag, then the error does bubble up.

What I'm saying is that any event condition that is even passed to event-cb *originates* from a cl-async callback. What this means is that any handler-bind that catches it must be *outside* the event loop. 

Keep in mind, that as of the last few commits, cl-async *does not catch errors inside of user-supplied callbacks by default*, they are allowed to bubble up. event-cb is used to notify the user only in cases where libuv gives any status other than success on a specific operation. For instance:

```lisp
(as:with-event-loop ()
  (as:with-delay ()
    (+ 4 'five)))
```

This triggers an error that passes all the way through cl-async to the top level/debugger, with full stack/restarts. In previous versions, the above error would be caught and rethrown inside of cl-async, erasing any stack or restarts.

That said, here's my understanding of what you want:

```lisp
;; cl-async's event-handling code
(catch 'event-handling-done
  (handler-bind
      ((error (lambda (e) (funcall event-cb e))))
    (error (make-instance 'tcp-eof :socket the-socket))))

;; event-cb
(lambda (e)
  (when (some-logic-here)
    (throw 'event-handling-done)))
```

Here's how it currently works (and the event-cb that would essentially do the same thing)

```lisp
;; cl-async's event-handling code
(funcall event-cb (make-instance 'tcp-eof :socket the-socket))

;; event-cb
(lambda (e)
  (unless (some-logic-here)
    (error e)))
```

I'm really not opposed to implementing the event handling from the first example and *by default* `throw`ing to `event-handling-done` once event-cb returns *unless* `cl-async:*throw-event-conditions*` (or something like it) is `t`. This would preserve backwards compatibility with the option of calling `error` on event conditions. But I do wonder if this is very useful because the error handling has to be done outside the event loop anyway, and the `(error some-event)` calls are going to share almost the exact same stack (originating in the cl-async error handling code).

> There is also the benefit that any restarts either cl-async's or mine established along the way can be called from any handler-bind

True, but calling `(error e)` from an event-cb would have the same effect here as throwing the event inside cl-async. I hope the examples above illustrate what I'm talking about.

>> The one exception I know of is the DNS error you found [...] I am open to changing this to another error type. 

> don't resort to catching, changing the error class and resignalling, use restarts instead.

What I meant was I'd be inclined to change the actual `(error "DNS error!")` call to something like `(error (make-instance 'event-error :msg "DNS error!"))`.

I hope that my days of catching and rethrowing errors are over =].

>> In fact, I'm starting to wonder if :catch-app-errors t is even useful, because it can now be completely mimicked by doing

> But I don't find it horrible, keep it.

Will do!

> Or, I'd make it a generalized boolean and get rid of :caught-errors entirely (or rather deprecate it). So if one passes nil to :catch-app-errors, everything bubbles up. If one passes t to it everything is handled through the as:handle-error generic function, which has some nice built-in methods that catch only cl-async errors and don't mess with user-specific errors at all. If one passes it a lambda, call that lambda with the error class. That function is resposible for handling all the errors. The user has the choice to add methods to as:handle-error or call it from his/her lambda.

I really like the idea of being able to call `:catch-app-errors` with a lambda and getting rid of `:caught-errors`.

> everything is handled through the as:handle-error generic function, which has some nice built-in methods that catch only cl-async errors and don't mess with user-specific errors at all

cl-async handles all these conditions internally. In other words, when it calls event-cb with a `socket-eof` condition, it already handles the closing of the socket and any needed cleanup at that point. I don't think there's a use case where having a generic event handler makes sense *inside* of cl-async, however that wouldn't stop the user from providing their own method for handling any errors that come through.

Actually though, now that I think about it, providing a generic function that gets called when `:catch-app-errors t` is passed would be kind of nice, and by default maybe it could just log any errors that came through.

I do like that idea.

---

So, for your final suggestions:

> make the fourth argument of functions such as as:tcp-connect optional. nil means everything bubbles up. t means use as:handle-error, a new generic and a lambda is a lambda as usual, but called with handler-bind semantics, i.e. if it doesn't throw or invoke restarts the error continues its journey.

As much as it pains me to change the `tcp-connect` API in such a large way, I think this makes the most sense. I think for now I will rename the `event-cb` arg in tcp-connect to `event-cb-dep` and add an `event-cb` keyword arg, and throw a deprecation warning if people pass a value to `event-cb-dep`. I'll keep this around for a few months until everyone has had enough time to update.

> add an exported catch tag of cl:event-handling-done alongside the cl-async-util:continue-event-loop that you apparently already provide. Export both.

I still think this may need a bit more discussion. If the event handling code examples I posted above are indeed what you envision, then I'm fine with implementing this, but am still curious as to what use it would be.

> Add a as:handle-error generic function and define some methods for it. Export that.

I like it.

> make the :catch-app-errors argument the same semantics as the previous item. it's only called when a specific request doesn't have an error handler attached, which is the point you made earlier. 

Actually my point (and I hope I made this clearer in the code examples above) was that event-cb doesn't catch user-generated errors at all anymore, it now only receives cl-async event conditions. One thing I've been kicking around is a `'send-to-eventcb` restart that would allow the user to send a caught error to the event-cb instead of what will be the function in `:catch-app-errors`. This preserves backward compatibility with the error handling of older cl-async versions. Perhaps this could even be a `*send-errors-to-eventcb*` variable to allow really easily switching to legacy error handling.

> error-catcher is deprecated.

Do you mean `:caught-errors`? If so, then yes it will have a very short life =].

**Comment by capitaomorte** [comment]
I will read the rest of you reply later, but right now, I just wanted to clarify this:

>> But why not let it have handler-bind semantics? I.e, if it doesn't invoke a restart or throw to some some exported cl:event-handling-done tag, then the error does bubble up.
> Keep in mind, that as of the last few commits, cl-async does not catch errors

I just wanted to clarify that I meant the *error handlers supplied by the user* should have "handler-bind semantics". I.e. when you detect that a user passed one, just `funcall` it inside your handler-bind handler that lets everything through. Is that what is done. I'll post an example later.

**Comment by orthecreedence** [comment]
I think i get it:

```lisp
;; current
(tcp-connect "somewhere.com" 80 'my-read-cb :event-cb 'do-something-with-error)

;; what I understand you are saying
(handler-bind
    ((error 'do-something-with-error))
  (tcp-connect "somewhere.com" 80 'my-read-cb))  ;; no :event-cb, handler is in handler-bind
```

Is that what you mean?

**Comment by capitaomorte** [comment]
@orthecreedence, no I don't think so. I'm saying something much simpler I think. I'm saying keep supporting `event-cb` args, but give them `handler-bind` *semantics*. This of course can mean a million of things and that's the source of the confusion I mean this.

Without `handler-bind` semantics, you probably now have.

```
;; error bubbles up
(tcp-connect "nowhere.com" 80 'my-read-cb)

;; Error does not bubble up
(tcp-connect "nowhere.com" 80 'my-read-cb :event-cb #'(lambda (e) (format t "oh noes~%")))
```

With `handler-bind` semantics, user's `:event-cb` has to explicitly find a point where control can be restarted, otherwise the error is untouched.

```
;; error bubbles up
(tcp-connect "nowhere.com" 80 'my-read-cb)

;; error bubbles up
(tcp-connect "nowhere.com" 80 'my-read-cb :event-cb #'(lambda () (format t "oh noes~%")))

;; error does not bubble up
(tcp-connect "nowhere.com" 80 'my-read-cb 
  :event-cb #'(lambda (e) (format t "oh noes~%") (invoke-restart cl-async:well-known-restart))

;; error does not bubble up
(tcp-connect "nowhere.com" 80 'my-read-cb 
  :event-cb #'(lambda (e) (format t "oh noes~%") (throw 'cl-async:well-known-tag nil))

;; error does not bubble up
(tcp-connect "nowhere.com" 80 'my-read-cb 
  :event-cb #'(lambda (e) (format t "oh noes~%") (return-from 'cl-async:well-known-block nil))
```

**Comment by capitaomorte** [comment]
But your previous example is also correct. I **also** mean that.

**Comment by orthecreedence** [comment]
Here's a quick example (using the latest version of cl-async):

```lisp
(handler-bind
    ((error
       (lambda (e)
         (let ((res (find-restart 'cl-async-util:continue-event-loop e)))
           (when res (invoke-restart res))))))
  (as:with-event-loop (:catch-app-errors nil
                       :send-errors-to-eventcb t)
    (as:delay
      (lambda () (format t "result: ~a~" (+ 4 'five)))
      :event-cb
        (lambda (e) (format t "got err: ~a~%" e)))))
```

I think what you're saying is you want something like this:

```lisp
(as:with-event-loop (:catch-app-errors nil
                     :send-errors-to-eventcb: t)
  (as:delay
    (lambda () (format t "result: ~a~%" (+ 4 'five)))
    :event-cb
      (lambda (e)
        (format t "got err: ~a~%" e)
        (throw 'cl-async-util:event-handling-done))))
```

The current version lets the error bubble to outside the loop, where the `continue-event-loop` restart sends it to the `:event-cb`. You're saying you want the error to be sent to the `:event-cb` *on its way up* with the option of doing a non-local exit to keep the error from bubbling?

I'm fairly certain this can be done in a backwards compatible manner with (yet another) switch somewhere (such as `:handle-error-in-eventcb` or something like it) which if `nil`, would just throw to the `event-handling-done` tag automatically after running the `event-cb`, otherwise if `t` would require that the `event-cb` throw to the tag if it wants to stop the error from bubbling. I guess the only challenge would be making sure events (`tcp-eof` etc) are handled the same way (although it might be possible to just wrap them in the same error handling code as everything else).
